### PR TITLE
Add a reader for MODIS Level 3 files in CMG format.

### PR DESCRIPTION
**Pull request by simonrp84**
This PR adds a reader for MODIS Level 3 (daily / monthly / etc) products on the climate monitoring grid.

I'd like feedback on the change I made to `hdfeos_base.py`. The CMG data has metadata containing URLs with `=` signs, which confuses the `_split_line` method. My change means that only the first `=` is used to split strings, with the subsequent ones assumed as part of the metadata.
Personally, I don't think the `try...except` is needed as there should only ever be one important `=` in each string. But I wrote it this way in case there's some problem I've not thought of. Opinions welcome.

 - [x] Tests added 
 - [x] Fully documented

### CODE REVIEW

**Comment by simonrp84** [comment]
Codefactor is moaning about the complexity of `create_hdfeos_test_file` and I'm not really sure what I'm supposed to do about it...all the if statements there are needed and it seems kinda wasteful to invent a whole new function just to avoid `if..then.else`. Suggestions appreciated!

File: satpy/readers/hdfeos_base.py [hunk 1]
```diff
@@ -148,7 +148,10 @@ def _read_mda(cls, lines, element=None):
 
     @classmethod
     def _split_line(cls, line, lines):
-        key, val = line.split("=")
+        try:
+            key, val = line.split("=")
+        except ValueError:
+            key, val = line.split("=", maxsplit=1)
```
Comment (djhoese) [inline satpy/readers/hdfeos_base.py:154]: I'd like @mraspaud to review this as I think he wrote the original version of this code.

File: satpy/tests/reader_tests/modis_tests/test_modis_l3.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    def test_load_l3_dataset(self, modis_l3_nasa_mod09_file):
+        """Load and check an L2 variable."""
+        scene = Scene(reader="modis_l3", filenames=modis_l3_nasa_mod09_file)
+
+        ds_name = "Coarse_Resolution_Surface_Reflectance_Band_2"
+        scene.load([ds_name])
+
+        data_arr = scene[ds_name]
+        assert isinstance(data_arr.data, da.Array)
+        data_arr = data_arr.compute()
```
Comment (djhoese) [inline satpy/tests/reader_tests/modis_tests/test_modis_l3.py:74]: Could you do me a favor and save the result to a new DataArray and verify that `data_arr.dtype == data_arr_np.dtype` and that `data_arr.dtype == np.float32` or whatever dtype you expect?

File: satpy/tests/reader_tests/modis_tests/_modis_fixtures.py [hunk 3]
```diff
@@ -251,8 +254,11 @@ def create_hdfeos_test_file(filename: str,
     if include_metadata:
         if geo_resolution is None or file_shortname is None:
             raise ValueError("'geo_resolution' and 'file_shortname' are required when including metadata.")
+        elif geo_resolution == -999 or geo_resolution == -9999:
+            setattr(h, 'StructMetadata.0', _create_struct_metadata_cmg(geo_resolution))  # noqa
+        else:
+            setattr(h, 'StructMetadata.0', _create_struct_metadata(geo_resolution))  # noqa
```
Comment (djhoese) [inline satpy/tests/reader_tests/modis_tests/_modis_fixtures.py:260]: How about extract this whole set of geo-related things to a `_add_geo_metadata` and calling it with `(h, geo_resolution)`?

File: satpy/readers/modis_l3.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        upperleft = self.metadata["GridStructure"]["GRID_1"]["UpperLeftPointMtrs"]
+        lowerright = self.metadata["GridStructure"]["GRID_1"]["LowerRightMtrs"]
+
+        # For some reason, a few of the CMG products multiply their
+        # decimal degree extents by one million. This fixes it.
+        if lowerright[0] > 1e6:
+            upperleft = tuple(val / 1e6 for val in upperleft)
+            lowerright = tuple(val / 1e6 for val in lowerright)
+
+        self.area_extent = (upperleft[0], lowerright[1], lowerright[0], upperleft[1])
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply satpy/readers/modis_l3.py:80]: I thought it was more efficient this way. Changed in new commit.

File: satpy/tests/reader_tests/modis_tests/test_modis_l3.py (hunk 2 above)
Comment (simonrp84) [reply satpy/tests/reader_tests/modis_tests/test_modis_l3.py:74]: Yes, that gives `True` and the dtype is `np.float32`

File: satpy/tests/reader_tests/modis_tests/_modis_fixtures.py (hunk 3 above)
Comment (simonrp84) [reply satpy/tests/reader_tests/modis_tests/_modis_fixtures.py:260]: Done!

File: satpy/readers/modis_l3.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            self.resolution = 360. / self.ncols
+        else:
+            self.resolution = float(gridname[pos:pos2])
+
+
+
+    def _sort_grid(self):
+        """Get the grid properties."""
+
+        # First, get the grid resolution
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply satpy/readers/modis_l3.py:66]: Fixed.

File: satpy/readers/modis_l3.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+"""
+import logging
+
+from pyresample import geometry
+
+from satpy.readers.hdfeos_base import HDFEOSGeoReader
+from satpy.utils import get_legacy_chunk_size
+
+logger = logging.getLogger(__name__)
+CHUNK_SIZE = get_legacy_chunk_size()
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply satpy/readers/modis_l3.py]: Removed.

File: satpy/tests/reader_tests/modis_tests/_modis_fixtures.py [hunk 7]
```diff
@@ -251,8 +254,11 @@ def create_hdfeos_test_file(filename: str,
     if include_metadata:
         if geo_resolution is None or file_shortname is None:
             raise ValueError("'geo_resolution' and 'file_shortname' are required when including metadata.")
+        elif geo_resolution == -999 or geo_resolution == -9999:
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply satpy/tests/reader_tests/modis_tests/_modis_fixtures.py]: Fixed.

File: satpy/tests/reader_tests/modis_tests/test_modis_l3.py (hunk 2 above)
Comment (djhoese) [reply satpy/tests/reader_tests/modis_tests/test_modis_l3.py:74]: Great. I'd like all new tests that compute some dask array to have something equivalent to this.

**Comment by djhoese** [comment]
CodeScene is happy now!

File: satpy/tests/reader_tests/modis_tests/test_modis_l3.py (hunk 2 above)
Comment (simonrp84) [reply satpy/tests/reader_tests/modis_tests/test_modis_l3.py]: Have now added this to the tests.

File: satpy/readers/hdfeos_base.py (hunk 1 above)
Comment (djhoese) [reply satpy/readers/hdfeos_base.py:154]: While I still want @mraspaud to review this change, @simonrp84 I'm confused why codecov says your new lines aren't covered. I suppose none of your tests check this actual metadata parsing?

File: satpy/readers/modis_l3.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# satpy.  If not, see <http://www.gnu.org/licenses/>.
+"""Modis level 3 hdf-eos format reader.
+
+Introduction
+------------
+
+The ``modis_l3`` reader reads Modis L3 products in hdf-eos format.
```
Comment (djhoese) [inline satpy/readers/modis_l3.py:23]:
```suggestion
The ``modis_l3`` reader reads MODIS L3 products in hdf-eos format.
```

File: satpy/readers/modis_l3.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        # First, get the grid resolution
+        self._get_res()
+
+        # Now compute the data extent
+        upperleft = self.metadata["GridStructure"]["GRID_1"]["UpperLeftPointMtrs"]
+        lowerright = self.metadata["GridStructure"]["GRID_1"]["LowerRightMtrs"]
+
+        # For some reason, a few of the CMG products multiply their
+        # decimal degree extents by one million. This fixes it.
+        if lowerright[0] > 1e6:
```
Comment (djhoese) [inline satpy/readers/modis_l3.py:74]: What are the extents known to be? Technically wouldn't this fail for any degrees between 0 and 1?

File: satpy/readers/modis_l3.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    def __init__(self, filename, filename_info, filetype_info, **kwargs):
+        """Init the file handler."""
+        super().__init__(filename, filename_info, filetype_info, **kwargs)
+
+        # Initialise number of rows and columns
+        self.nrows = self.metadata["GridStructure"]["GRID_1"]["YDim"]
+        self.ncols = self.metadata["GridStructure"]["GRID_1"]["XDim"]
+
+        # Get the grid name and other projection info
+        self.area_extent = self._sort_grid()
```
Comment (djhoese) [inline satpy/readers/modis_l3.py:90]: Could you please move this to the top of all the methods? I find it very hard to understand what a class is doing when the `__init__` is not the first method. You could also do "newspaper" style ordering of the methods and do `__init__` -> `_sort_grid` -> `_get_res`.

File: satpy/readers/modis_l3.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+
+    def available_datasets(self, configured_datasets=None):
+        """Automatically determine datasets provided by this file."""
+        logger.debug("Available_datasets begin...")
+
+        ds_dict = self.sd.datasets()
+
+        yield from super().available_datasets(configured_datasets)
+        common = {"file_type": "mcd43_cmg_hdf", "resolution": self.resolution}
```
Comment (djhoese) [inline satpy/readers/modis_l3.py:100]: I'm not sure the debug log message is providing any useful information to the user or a log file. I also reordered when the `self.sd_datasets` is accessed since it isn't used until after the `yield from`.

```suggestion
        yield from super().available_datasets(configured_datasets)
        common = {"file_type": "mcd43_cmg_hdf", "resolution": self.resolution}
        ds_dict = self.sd.datasets()
```

File: satpy/readers/modis_l3.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+        self.area_extent = self._sort_grid()
+
+
+    def available_datasets(self, configured_datasets=None):
+        """Automatically determine datasets provided by this file."""
+        logger.debug("Available_datasets begin...")
+
+        ds_dict = self.sd.datasets()
+
+        yield from super().available_datasets(configured_datasets)
```
Comment (djhoese) [inline satpy/readers/modis_l3.py:99]: I don't think you can blindly yield from this unless you make two assumptions:

1. There will never be more than one file passed to the reader at a time.
2. There will never be a dataset configured in the YAML.

For 1, you are likely redefining every dataset that is in every file since this `yield from` will produce all the datasets from other file handlers. If you have two files passed to the reader (meaning two file handlers are created) and they had the same variables then you'd get the first set from the first file handler which would then be yielded from this `yield from` line, then the same variables are yielded again in the below for loop. If two files with the same variables are never read at the same time then maybe this isn't a problem, but it seems like a future "gotcha".

For 2, if there is a chance that the variable (file_key) defined in the YAML isn't in the file being read then instead of a `yield` from, we should maybe be checking that the filetype matches this current file handler and that the variable exists in `ds_dict`. Also, if it does exist in `ds_dict` (it is part of the file) then we may not want to yield a copy of it in the below for loop.

I typically keep a `set()` object around of variable names that I've seen already and don't yield them again.

File: satpy/readers/modis_l3.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        area = geometry.AreaDefinition("gridded_modis",
+                                       "A gridded L3 MODIS area",
+                                       "longlat",
+                                       proj_param,
+                                       self.ncols,
+                                       self.nrows,
+                                       self.area_extent)
+        self.area = area
+
+        return self.area
```
Comment (djhoese) [inline satpy/readers/modis_l3.py:132]:
```suggestion
                                       self.area_extent)
        return area
```

Comment (djhoese) [reply satpy/readers/modis_l3.py:132]: I'd also really prefer if `self.area_extent` didn't have to be an instance attribute. If the main reason is so we can have the error from `_get_res` about only CMG grids being supported and have it raised on file handler creation, then fine. But otherwise, doing those checks on `__init__` is only slowing down Scene creation (admittedly probably not much) when datasets may not be loaded from this file.

File: satpy/tests/reader_tests/modis_tests/test_modis_l3.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+    def test_available_reader(self):
+        """Test that MODIS L3 reader is available."""
+        assert "modis_l3" in available_readers()
+
+    @pytest.mark.parametrize(
+        ("loadable", "filename"),
+        [
+            ("Coarse_Resolution_Surface_Reflectance_Band_2", lazy_fixture("modis_l3_nasa_mod09_file")),
+            ("BRDF_Albedo_Parameter1_Band2",lazy_fixture("modis_l3_nasa_mod43_file")),
```
Comment (djhoese) [inline satpy/tests/reader_tests/modis_tests/test_modis_l3.py:56]: @mraspaud why isn't ruff complaining about the lack of a space after the comma?

**Review by djhoese — CHANGES_REQUESTED** [review]
Nice job putting this together. I put a couple questions, some concerns, and other comments inline.

Otherwise, I'd like to officially state that I'd like something better than the -999/-9999 resolution stuff in the tests to describe these files. It just seems so hacky. If it must be this way then fine, but it feels like we can come up with something better.

File: satpy/readers/modis_l3.py (hunk 13 above)
Comment (simonrp84) [reply satpy/readers/modis_l3.py:132]: Aah, I asked about this on slack (https://pytroll.slack.com/archives/C0LNH7LMB/p1699437836789399) but misunderstood your reply. I'll remove the code.

File: satpy/readers/hdfeos_base.py (hunk 1 above)
Comment (simonrp84) [reply satpy/readers/hdfeos_base.py:154]: I've updated the tests so that both the old and new lines are covered.

File: satpy/readers/modis_l3.py (hunk 8 above)
Comment (simonrp84) [reply satpy/readers/modis_l3.py:23]: Fixed.

File: satpy/readers/modis_l3.py (hunk 9 above)
Comment (simonrp84) [reply satpy/readers/modis_l3.py:74]: Extents should always be -180 -> 180 longitude. Can't see why this would fail for values between 0, 1 as those are considerably smaller than the `1e6` value being tested here.

File: satpy/readers/modis_l3.py (hunk 10 above)
Comment (simonrp84) [reply satpy/readers/modis_l3.py:90]: Done!

File: satpy/readers/modis_l3.py (hunk 9 above)
Comment (djhoese) [reply satpy/readers/modis_l3.py:74]: If the grid of the data is always near-global then this isn't an issue. If the right-most side of the data is at 0 degrees longitude (or up to 1 degrees longitude) then 0 * 1e6 is not `> 1e6` so your check will not catch this and that it needs to divide by 1e6. Same thing happens for degrees of longitude up to 1 where 1 * 1e6 is also not `> 1e6`.

File: satpy/readers/modis_l3.py (hunk 11 above)
Comment (simonrp84) [reply satpy/readers/modis_l3.py:100]: Changed to this (will refactor more once I sort out this dynamic datasets stuff.

File: satpy/tests/reader_tests/modis_tests/test_modis_l3.py (hunk 14 above)
Comment (simonrp84) [reply satpy/tests/reader_tests/modis_tests/test_modis_l3.py:56]: Space is there now :)

File: satpy/readers/modis_l3.py (hunk 9 above)
Comment (simonrp84) [reply satpy/readers/modis_l3.py:71]: Oh, of course. This should never happen but just to be sure I'll also add a check for the uppermost longitude value.

File: satpy/readers/modis_l3.py (hunk 12 above)
Comment (simonrp84) [reply satpy/readers/modis_l3.py:101]: I've tried to address this in a new commit now - hopefully it satisfies what you're requesting.
Coverage is poor right now as `configured_datasets` is empty. Will try to fix 'em.

Comment (djhoese) [reply satpy/readers/modis_l3.py:101]: I'm still reading the rest of this, but you'll want to remove this `yield from` line as you're handling `configured_datasets` below.

Comment (simonrp84) [reply satpy/readers/modis_l3.py:101]: Hah yes, I was just debugging that now.

File: satpy/readers/modis_l3.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+            handled_var_names.add(file_key)
+            if is_avail is not None:
+                # some other file handler said it has this dataset
+                # we don't know any more information than the previous
+                # file handler so let's yield early
+                yield is_avail, ds_info
+                continue
+            if self.file_type_matches(ds_info["file_type"]) is None:
+                # this is not the file type for this dataset
+                yield None, ds_info
```
Comment (djhoese) [inline satpy/readers/modis_l3.py:120]: I think this is a bug in the `viirs_edr.py` reader, but I think this should be followed by a `continue`. The reason being that if we aren't the file for this data set (file type doesn't match) then we shouldn't be claiming anything about it (the below `yield file_key in ds_dict.keys(), ds_info` line is doing this).

Unless you disagree, could you put a `continue` below this yield and fix the `viirs_edr.py` reader? To trigger this bug you'd have to process a file type that did not match one of the YAML-defined datasets, but even then the end result for the user would be the same (the variable isn't available). :thinking: I guess if you loaded multiple file types and only one of them matched the YAML-defined variable and this...no I think the behavior would be the same.

File: satpy/readers/modis_l3.py [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+            yield file_key in ds_dict.keys(), ds_info
+
+        yield from self._dynamic_variables_from_file(handled_var_names)
+
+    def _dynamic_variables_from_file(self, handled_var_names: set) -> Iterable[tuple[bool, dict]]:
+
```
Comment (djhoese) [inline satpy/readers/modis_l3.py:126]:
```suggestion
```

File: satpy/readers/modis_l3.py [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+        for var_name in self.sd.datasets().keys():
+            if var_name in handled_var_names:
+                # skip variables that YAML had configured, but allow lon/lats
+                # to be reprocessed due to our dynamic coordinate naming
```
Comment (djhoese) [inline satpy/readers/modis_l3.py:130]:
```suggestion
                # skip variables that YAML had configured
```

File: satpy/readers/modis_l3.py [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+class ModisL3GriddedHDFFileHandler(HDFEOSGeoReader):
+    """File handler for MODIS HDF-EOS Level 3 CMG gridded files."""
+
+    def __init__(self, filename, filename_info, filetype_info, **kwargs):
+        """Init the file handler."""
+        super().__init__(filename, filename_info, filetype_info, **kwargs)
+
+        # Initialise number of rows and columns
+        self.nrows = self.metadata["GridStructure"]["GRID_1"]["YDim"]
+        self.ncols = self.metadata["GridStructure"]["GRID_1"]["XDim"]
```
Comment (djhoese) [inline satpy/readers/modis_l3.py:54]: Correct me if I'm wrong, but these aren't used except in getting the area. I don't see a reason for this to happen in `__init__`. Could you move the access of these into the area creation method (and not assign them to `self`)? If you expect this to error out to notify the user of a problem with there file then I guess maybe this is fine, but should probably have an explicit error if that's the case.

File: satpy/readers/modis_l3.py (hunk 15 above)
Comment (simonrp84) [reply satpy/readers/modis_l3.py:120]: I just asked about this on slack as my tests brought up this problem (the same DS was being returned twice, one with `None` and one with `False`). Will add in the `continue`

File: satpy/readers/modis_l3.py [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+
+        # Get the grid name and other projection info
+        self.area_extent = self._sort_grid()
+
+
+    def _sort_grid(self):
+        """Get the grid properties."""
+        # First, get the grid resolution
+        self._get_res()
```
Comment (djhoese) [inline satpy/readers/modis_l3.py:63]: Please move this line to the `__init__` and change it to `self.resolution = self._get_res()`. This avoids the hiding of instance variables inside methods called from init. Especially in this case where it is a method called inside a method called from init.

File: satpy/readers/modis_l3.py (hunk 18 above)
Comment (simonrp84) [reply satpy/readers/modis_l3.py:54]: done!

File: satpy/readers/modis_l3.py [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+            yield True, common
+
+    def get_dataset(self, dataset_id, dataset_info):
+        """Get DataArray for specified dataset."""
+        dataset_name = dataset_id["name"]
+        dataset = self.load_dataset(dataset_name, dataset_info.pop("category", False))
+        self._add_satpy_metadata(dataset_id, dataset)
+
+        return dataset
+
```
Comment (djhoese) [inline satpy/readers/modis_l3.py:143]:
```suggestion
```

File: satpy/readers/modis_l3.py (hunk 19 above)
Comment (simonrp84) [reply satpy/readers/modis_l3.py:63]: done!

File: satpy/tests/reader_tests/modis_tests/_modis_fixtures.py [hunk 21]
```diff
@@ -326,8 +339,35 @@ def _create_struct_metadata(geo_resolution: int) -> str:
...
+                              f"{upright}\n"
+                              "END_GROUP=GRID_1\n"
+                              "END_GROUP=GridStructure\nEND")
+    return struct_metadata_header
+
+
 def _create_header_metadata() -> str:
-    archive_metadata_header = "GROUP = ARCHIVEDMETADATA\nEND_GROUP = ARCHIVEDMETADATA\nEND"
+    archive_metadata_header = ("GROUP = ARCHIVEDMETADATA\n"
+                               'TEST_URL = "http://modis.gsfc.nasa.gov/?some_val=100"\n'
```
Comment (djhoese) [inline satpy/tests/reader_tests/modis_tests/_modis_fixtures.py:369]: Does this work? For consistency of quote type.

```suggestion
                               "TEST_URL = \"http://modis.gsfc.nasa.gov/?some_val=100\"\n"
```

File: satpy/readers/modis_l3.py (hunk 20 above)
Comment (simonrp84) [reply satpy/readers/modis_l3.py:120]: Done!

File: satpy/readers/modis_l3.py (hunk 17 above)
Comment (simonrp84) [reply satpy/readers/modis_l3.py]: Oops, got lazy with copy-paste. Fixed.

File: satpy/tests/reader_tests/modis_tests/_modis_fixtures.py (hunk 21 above)
Comment (simonrp84) [reply satpy/tests/reader_tests/modis_tests/_modis_fixtures.py:365]: Ruff complains if I do this.

File: satpy/tests/reader_tests/modis_tests/_modis_fixtures.py [hunk 22]
```diff
@@ -228,9 +228,7 @@ def generate_imapp_filename(suffix):
 
 def create_hdfeos_test_file(filename: str,
                             variable_infos: dict,
-                            geo_resolution: Optional[int] = None,
-                            file_shortname: Optional[str] = None,
-                            include_metadata: bool = True):
+                            metadata_dict: Optional[dict] = {}):
```
Comment (djhoese) [inline satpy/tests/reader_tests/modis_tests/_modis_fixtures.py:231]: No mutable default arguments please.

```suggestion
                            metadata_dict: Optional[dict] = None):
```

File: satpy/tests/reader_tests/modis_tests/_modis_fixtures.py [hunk 23]
```diff
@@ -243,17 +241,27 @@ def create_hdfeos_test_file(filename: str,
...
     """
     h = SD(filename, SDC.WRITE | SDC.CREATE)
 
-    if include_metadata:
-        if geo_resolution is None or file_shortname is None:
-            raise ValueError("'geo_resolution' and 'file_shortname' are required when including metadata.")
-        setattr(h, 'CoreMetadata.0', _create_core_metadata(file_shortname))  # noqa
-        setattr(h, 'StructMetadata.0', _create_struct_metadata(geo_resolution))  # noqa
-        setattr(h, 'ArchiveMetadata.0', _create_header_metadata())  # noqa
+    if metadata_dict is not None and metadata_dict != {}:
```
Comment (djhoese) [inline satpy/tests/reader_tests/modis_tests/_modis_fixtures.py:249]: I'm not sure this has to be optional at all. I wasn't thinking it'd be the file_shortname and stuff, it'd be instead:


```
metadata = {
    "StructMetadata.0": _create_struct_metadata(geo_res),
    "CoreMetadata.0": _create_core_metadata(file_shortname),
    "ArchiveMetadata.0": _create_header_metadata(),
}
create_hdfeos_test_file(filename, variable_infos, metadata)
```

And `metadata` would be a positional argument (not optional) and could be passed as `{}` for any files that need no metadata.

A whole other idea would be to make the struct/geo metadata, core metadata, and archive metadata individual keyword arguments. That way the user/caller doesn't need to know the exact "StructMetadata.0" attribute name which they could easily screw up.

File: satpy/tests/reader_tests/modis_tests/_modis_fixtures.py [hunk 24]
```diff
@@ -326,8 +334,36 @@ def _create_struct_metadata(geo_resolution: int) -> str:
...
+    # Case of a MOD09 file
+    if ftype == "MOD09":
+        gridline = 'GridName="MOD09CMG"\n'
+        upleft = "UpperLeftPointMtrs=(-180000000.000000,90000000.000000)\n"
+        upright = "LowerRightMtrs=(180000000.000000,-90000000.000000)\n"
+    # Case of a MCD43 file
+    elif ftype == "MCD43C1":
+        gridline = 'GridName="MCD_CMG_BRDF_0.05Deg"\n'
+        upleft = "UpperLeftPointMtrs=(-180.000000,90.000000)\n"
+        upright = "LowerRightMtrs=(180.000000,-90.000000)\n"
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply satpy/tests/reader_tests/modis_tests/_modis_fixtures.py]: Sorted

File: satpy/readers/modis_l3.py [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+class ModisL3GriddedHDFFileHandler(HDFEOSGeoReader):
+    """File handler for MODIS HDF-EOS Level 3 CMG gridded files."""
+
+    def __init__(self, filename, filename_info, filetype_info, **kwargs):
+        """Init the file handler."""
+        super().__init__(filename, filename_info, filetype_info, **kwargs)
+
+        # Get the grid resolution, name and other projection info
+        self.resolution = self._get_res()
```
Comment (djhoese) [inline satpy/readers/modis_l3.py:53]: Ok, don't hate me, but this is only used in the available datasets, right? Could you just get this right above the for loop:

```
res = self._get_res()
for var_name in self.sd.datasets().keys():
```

(and move `_get_res` to just below the dynamic available datasets method)

File: satpy/tests/reader_tests/modis_tests/_modis_fixtures.py [hunk 26]
```diff
@@ -228,32 +227,29 @@ def generate_imapp_filename(suffix):
...
-            raise ValueError("'geo_resolution' and 'file_shortname' are required when including metadata.")
-        setattr(h, 'CoreMetadata.0', _create_core_metadata(file_shortname))  # noqa
-        setattr(h, 'StructMetadata.0', _create_struct_metadata(geo_resolution))  # noqa
-        setattr(h, 'ArchiveMetadata.0', _create_header_metadata())  # noqa
+    if struct_meta != "":
+        setattr(h, "StructMetadata.0", struct_meta)
+    if core_meta != "":
+        setattr(h, "CoreMetadata.0", core_meta)
+    if archive_meta != "":
+        setattr(h, "ArchiveMetadata.0", archive_meta)
```
[the remark this answers is not in the document]
Comment (simonrp84) [reply satpy/tests/reader_tests/modis_tests/_modis_fixtures.py:252]: Changed.

File: satpy/readers/modis_l3.py (hunk 25 above)
Comment (simonrp84) [reply satpy/readers/modis_l3.py]: Done!

**Review by djhoese — APPROVED** [review]
This looks good to me. I wouldn't mind a discussion on the usage of EPSG:4326 and I'd still like @mraspaud to look at the change to hdfeos_base.py, but otherwise this looks pretty good. Nice job sticking with it @simonrp84!

File: satpy/readers/hdfeos_base.py (hunk 1 above)
Comment (mraspaud) [reply satpy/readers/hdfeos_base.py:154]: I'm good with this, but why not just use the maxsplit argument everything? It might even make this slightly faster.

File: satpy/readers/modis_l3.py [hunk 27]
```diff
@@ -0,0 +1,137 @@
...
+        return dataset
+
+
+    def get_area_def(self, dsid):
+        """Get the area definition.
+
+        This is fixed, but not defined in the file. So we must
+        generate it ourselves with some assumptions.
+        """
+        proj_param = "EPSG:4326"
```
[the remark this answers is not in the document]
Comment (djhoese) [reply satpy/readers/modis_l3.py:134]: I think your suggestion on slack to use `OGC:CRS84` is good as a minimum. I could have sworn there was a CRS that was at least rooted in an exact date/time, but otherwise was similar to EPSG 4326. At least CRS84 has the same axis order that we use/expect.

File: satpy/readers/hdfeos_base.py (hunk 1 above)
Comment (djhoese) [reply satpy/readers/hdfeos_base.py:154]: You mean remove the try/except and only use the `maxsplit=1` version of the call? Yeah I guess that was the question, but seems obvious now. Without maxsplit=1, this will only ever work if there is a single `"="` to split on so always specifying it makes the most sense.

File: satpy/readers/viirs_edr.py [hunk 28]
```diff
@@ -215,6 +215,7 @@ def available_datasets(self, configured_datasets=None):
             if self.file_type_matches(ds_info["file_type"]) is None:
                 # this is not the file type for this dataset
                 yield None, ds_info
+                continue
```
Comment (mraspaud) [inline satpy/readers/viirs_edr.py:218]: I'm probably missing something, but what does this have to do with modis l3?

**Comment by simonrp84** [comment]
Looking through the MODIS docs it seems like `EPSG:4326` is what's used by the data provider, so I suggest that we stick with that in this PR.

**Review by mraspaud — APPROVED** [review]
Looks good, but I would use the maxsplit everytime.

File: satpy/readers/viirs_edr.py (hunk 28 above)
Comment (simonrp84) [reply satpy/readers/viirs_edr.py:218]: Nothing, but we noticed that bug while doing this PR - so Dave suggested I fix it.
[i.e: I'm blaming Dave ;-) ]

File: satpy/readers/hdfeos_base.py (hunk 1 above)
Comment (simonrp84) [reply satpy/readers/hdfeos_base.py:154]: OK, will change.

Comment (simonrp84) [reply satpy/readers/hdfeos_base.py]: Done!

**Comment by simonrp84** [comment]
> Looks good, but I would use the maxsplit everytime.

Sorted this, so always using `maxsplit=1` now.

File: satpy/readers/viirs_edr.py (hunk 28 above)
Comment (djhoese) [reply satpy/readers/viirs_edr.py:218]: Technically the VIIRS EDR reader should probably have a test for this. I could add to this PR and copy what @simonrp84 did for his test of `available_datasets` since that seems like one of the only ways to verify this (as it does generally depend on the order of file types being processed and what's in the YAML file).

**Comment by djhoese** [comment]
I'll see if I can do a quick check on the viirs edr reader so complete the tests for these changes. Then we can merge it.

**Comment by simonrp84** [comment]
That was a monster of a PR, turned out to be more complex than I was anticipating! Thanks for the help and comments along the way, really nice to get this merged.
